docs/adr/adr_003_bundling_collection_data.md

```diff
@@ -0,0 +1,209 @@
+
+In future calls of `download()` on records whose attachment has been modified on the server, the client would fetch them individually as it does with the current code. In other words, in this ADR, we don’t address building “delta” or “diff” bundles for clients that already have local data.
+
+Consumers would choose to explicitly use this new method in their code, based on their use-case. And we (SysEng) could easily modify existing code to rewrite the code of some use-cases that rely on massive parallel fetch (eg. replace with a call to `cacheAll()`).
+
+Requirements
+- Total amount of time to download, extract, and process, must not take longer than fetching attachments individually
+- Handling the bundle in the client should not bloat the main thread or impact user experience in general
+- Feature must not get in the way of legacy clients
+- Implementation should be resilient to race conditions on attachment downloads. Unpacking the bundle can overwrite attachments that had been previously downloaded individually, and individual downloads can overwrite unbundled ones
+- Bundle format should be easily readable by client (third-party libraries are hard to ship in Gecko)
+
+References
+- [The intermediates download code](https://searchfox.org/mozilla-central/rev/8ec3cc0472ad4f51b254728d024b696eaba82ba0/security/manager/ssl/RemoteSecuritySettings.sys.mjs#335-365) that fetches attachments in bulk, and stores state in records
+- [Download of crlite files](https://searchfox.org/mozilla-central/rev/8ec3cc0472ad4f51b254728d024b696eaba82ba0/security/manager/ssl/RemoteSecuritySettings.sys.mjs#620-637)
+- [Download of all regions files](https://searchfox.org/mozilla-central/rev/8ec3cc0472ad4f51b254728d024b696eaba82ba0/security/manager/ssl/RemoteSecuritySettings.sys.mjs#335-365)
+- [Download of all addons suggestions files](https://searchfox.org/mozilla-central/rev/8ec3cc0472ad4f51b254728d024b696eaba82ba0/browser/components/urlbar/private/AddonSuggestions.sys.mjs#87-101)
+
+### Metrics To Measure Before/After
+We want to measure CPU usage, total bandwidth used, duration of syncronization on fresh profiles, and battery usage for full synchronization. We should be able to gather most of this by running power and performance profiles on different hardware.
+
+Hardware to include:
+- Mac Desktop/Laptop
+- Linux Desktop/Laptop
+- Windows Desktop/Laptop
+- Android Mobile
+- iOS Mobile (does it use remote-settings?)
+
+
+## Non Goals
+- Support diffs or delta bundles for clients that already have local data
+- Reduce client performance (ex: we shouldn't increase power usage for clients)
+
+## Assumptions
+- We will find a way for Gecko clients to extract at least one archive format (very likely using third-party library)
+
+## Decision Drivers
```

Comment:
We mention whether the solution is generic (i.e. could be included in `kinto-attachment`) or Remote Settings specific a few times in the options below. Should we mention that as a decision driver here?

Reply:
Adjusting these to be a list and mentioned that